src/store/store.js

```diff
@@ -0,0 +1,265 @@
+			listener(null, error);
+		});
+	}
+
+	setError(entityId, token, error) {
+		const lowerCaseEntityId = entityId.toLowerCase();
+		const tokenCache = token.toString();
+		this._entities.get(tokenCache).set(lowerCaseEntityId, {
+			status: 'error',
+			entity: null,
+			error: error,
+			request: null
+		});
+		this.notifyError(entityId, token, error);
+	}
+
+	_setFetchingState(entityId, token) {
+		const lowerCaseEntityId = entityId.toLowerCase();
+		const tokenCache = token.toString();
+		this._entities.get(tokenCache).set(lowerCaseEntityId, {
+			status: 'fetching',
+			entity: null
+		});
+	}
+
+	_initContainer(map, entityId, token, init) {
+		const lowerCaseEntityId = entityId.toLowerCase();
+		const tokenCache = token.toString();
+		if (!map.has(tokenCache)) {
+			map.set(tokenCache, new Map());
+		}
+		const entityMap = map.get(tokenCache);
+		if (init && !entityMap.has(lowerCaseEntityId)) {
+			entityMap.set(lowerCaseEntityId, init);
+		}
+		return entityMap.get(lowerCaseEntityId);
+	}
+
+	async _handleCachePriming(token, response) {
+		const linkHeaderValues = response.headers && response.headers.get('Link');
+		if (!linkHeaderValues) {
+			return;
+		}
+
+		const cachePrimers = this._parseLinkHeader(linkHeaderValues)
+			.filter(function(link) {
+				return link.rel.indexOf('https://api.brightspace.com/rels/cache-primer') !== -1;
+			});
+
+		if (cachePrimers.length === 0) {
+			return;
+		}
+
+		await Promise.all(cachePrimers.map((cachePrimer) => {
+			return this.fetch(cachePrimer.href, token, true);
+		}));
+	}
+
+	// parse a Link header
+	//
+	// Link:<https://example.org/.meta>; rel=meta
+	//
+	// var r = parseLinkHeader(xhr.getResponseHeader('Link');
+	// r['meta'] outputs https://example.org/.meta
+	//
+	_parseLinkHeader(links) {
+		const linkexp = /<[^>]*>\s*(\s*;\s*[^\(\)<>@,;:"\/\[\]\?={} \t]+=(([^\(\)<>@,;:"\/\[\]\?={} \t]+)|("[^"]*")))*(,|$)/g; // eslint-disable-line no-useless-escape
+		const paramexp = /[^\(\)<>@,;:"\/\[\]\?={} \t]+=(([^\(\)<>@,;:"\/\[\]\?={} \t]+)|("[^"]*"))/g; // eslint-disable-line no-useless-escape
+
+		const matches = links.match(linkexp);
+		const _links = [];
+		for (let i = 0; i < matches.length; i++) {
+			const split = matches[i].split('>');
+			const href = split[0].substring(1);
+			_links.push({
+				href
+			});
+			const ps = split[1];
+			const s = ps.match(paramexp);
+			for (let j = 0; j < s.length; j++) {
+				const p = s[j];
+				const paramsplit = p.split('=');
+				const name = paramsplit[0];
+				const val = paramsplit[1].replace(/["']/g, '');
+				if (name === 'rel') {
+					const relsplit = val.split(' ');
+					_links[i][name] = relsplit;
+				} else {
+					_links[i][name] = val;
+				}
+			}
+		}
+		return _links;
+	}
+}
+
+window.D2L = window.D2L || {};
+window.D2L.SirenSdk = window.D2L.SirenSdk || {};
+window.D2L.SirenSdk.EntityStore = window.D2L.SirenSdk.EntityStore || new EntityStore(window.d2lfetch);
+
+export async function fetch(entityIds, token, bypassCache) {
```

Comment:
The design pattern for this is keeping the entity store class more lean and focused. So I am following a more functional and object approach. I also feel like these methods have more knowledge on how things are stored and such so splitting them from the class makes sense to me. 

For es6 modules, this method actually allows private code which is also quite the bonus. It also expressing the code what is intended to be used by the consumer.

For this specifically, I didn't think the class should care about how to resolve a token. It should only know how to work with a resolved token. This allows for the method within the class to use each other without resolving the token over and over again. This is great because resolving a token requires the method to be async which isn't always the desired behaviour. Such as removing a listener.

However, a consumer shouldn't have direct access to a resolved token since they can change. So the methods that are exported are all async and resolve the token. (Note for removing a listener the `addListener` method returns a function that removes the listener with an already resolved token. So this is how the user can take advantage of a resolved token without having access to it.)

With this approach it also allows us to make a batch call such that we only resolve the token once. This might be over optimization creep though. It does make it easier to work with multiple sub-entities though :D

Reply:
>The design pattern for this is keeping the entity store class more lean and focused. So I am following a more functional and object approach. I also feel like these methods have more knowledge on how things are stored and such so splitting them from the class makes sense to me.

Respectfully, i disagree.  I personally don't see anything wrong with have differing function signatures attached to the store.  It's not like you could use those functions without the entity store, so you could never take advantage of something like tree shaking or bundles. 

Additionally if you are sitting inside of an ide or some sort of smart text editor, the discoverability of the bulk actions only exists if you're inspecting the exports instead of the store itself which could be a disadvantage.

>For this specifically, I didn't think the class should care about how to resolve a token. It should only know how to work with a resolved token. This allows for the method within the class to use each other without resolving the token over and over again. This is great because resolving a token requires the method to be async which isn't always the desired behaviour. Such as removing a listener.

Not sure i really follow, You've already injected the getToken method into the file so using it inside the class doesn't seem like an issue.  You can mix async methods in for where it makes sense, and even provide async wrappers around methods you want to be synchronous.

Reply:
Your suggestion is to embed the global variables within the class and each instance of the class would just access them? Instead of global variable being hidden and is the class?

I still would like to see the store used in NodeJs applications as well. So the separation above brings us closer to that. I could separate these methods out into their own file. Since this is the only use case I figured why expose the class at this moment.

I also like importing the methods you need vs the entire class.

Reply:
> I also like importing the methods you need vs the entire class.

Is that just because you can't have private members in a class so you're using module exports to achieve the same results?

Reply:
This may be the underline reason since I really like a clean interface. Here the class is implementation detail that doesn't need to be exposed. The store only requires the methods we are exporting and the class just makes implementation of those easier.

Exporting the class here seems over kill and making the interface complicated. This will then make it harder to maintain and so on.